internal/json/json_unmarshal.go

```diff
@@ -21,11 +21,19 @@ func (s *scopeJSON) ToPrincipalResourceNode() (isPrincipalResourceScopeNode, err
 	switch s.Op {
 	case "All":
 		return ast.Scope{}.All(), nil
-	case "==":
-		if s.Entity == nil {
-			return nil, fmt.Errorf("missing entity")
+	case "==": //todo: wip: change other operators and nodes to use this pattern
+		var ref types.EntityReference
+
+		switch {
+		case s.Slot != nil:
+			ref = types.VariableSlot{Name: types.String(*s.Slot)}
+		case s.Entity != nil:
+			ref = types.EntityUID(*s.Entity)
```

Comment:
We may have to imitate Rust implementation and have slots in the `Policy` type.

The following policy is valid:
```cedar
@id("AccessVacation")
permit(
    principal,
    action == Action::"view",
    resource
);


@id("Template")
permit(
    principal == ?principal,
    action == Action::"view",
    resource
);
```

It contains both a static policy and a template. Because of that, in Rust, a policy is defined as:
```rust
pub struct Policy {
    /// Reference to the template
    template: Arc<Template>,
    /// Id of this link
    /// None in the case that this is an instance of a Static Policy
    link: Option<PolicyID>,
    // INVARIANT (values total map)
    // All of the slots in `template` MUST be bound by `values`
    /// values the slots are bound to.
    /// The constructor `new` is only visible in this module,
    /// so it is the responsibility of callers to maintain
    values: HashMap<SlotId, EntityUID>,
}
```


I started splitting the serialization so that slots could only be parsed by a `type Template struct`. But because the above policy is valid, we couldn't know ahead of time whether to use a `Policy` or `Template`. 

Unless we create a `cedarDTO` that could be either policy or template and decide which type to use based on whether there are slots, but I wouldn't say I like this idea.

However, instead of doing exactly like Rust, I think we could compromise to avoid a massive change in the codebase by changing `ast.Policy` to:
```go
type templateContext struct {
    linkID types.String
    slots  map[types.SlotID]types.EntityUID
}

type Policy struct {
    Effect      Effect
    Annotations []AnnotationType // duplicate keys are prevented via the builders
    Principal   IsPrincipalScopeNode
    Action      IsActionScopeNode
    Resource    IsResourceScopeNode
    Conditions  []ConditionType
    Position    Position

    templateContext templateContext
}
```

P.S. I'm still trying to think of a better name than `templateContext`. The case is to encapsulate template-related information in a new field instead of encapsulating the expression fields in a `Template` field, which would demand changes in all the codebase.

Reply:
Hi @patjakdev, do you have any insights on this?